### DISCUSSION TITLE
no space left on device for prometheus

### DISCUSSION THREAD

**Discussion by ShurikAg** [digitalocean/Kubernetes-Starter-Kit-Developers#165] [Q&A] [answered]
Hi all, 

We are running on this stack for a while now. After some time I noticed that we ran out of space on prometheus storage.

`ts=2022-06-01T07:36:18.482Z caller=main.go:1064 level=error err="opening storage failed: open /prometheus/wal/00000554: no space left on device"`

The storage configuration is exactly as it was proposed from the [starter kit](https://github.com/digitalocean/Kubernetes-Starter-Kit-Developers/tree/main/04-setup-prometheus-stack#step-5---configuring-persistent-storage-for-prometheus). So, it is set to 5G.

Now, I tried adjusting the storage size to 20G through the same repo that drives the cluster config, however, the storage still stays at 5G. How can I resize the storage allocated for it?

**Comment by v-ctiutiu** [comment, marked as answer]
Hi @ShurikAg 

Let me check and get back to you.

Thanks.

**Reply by ShurikAg** [reply]
I feel that I broke some things. I deleted the storage class by mistake. But, it seems like nothing really changed.
As I mentioned before, updating the storage size through flux didn't do anything. So, I decided to update it directly in Kubernetes dashboard. 
<img width="1250" alt="image" src="https://user-images.githubusercontent.com/916432/171805584-c2b668fa-2192-461b-aca9-8c2475124166.png">

Now, pushing anything on this cluster infra, the prometheus is failing to restart 
<img width="1192" alt="image" src="https://user-images.githubusercontent.com/916432/171817475-4f01639c-07e8-4cb4-b3ef-cfe3ea90dd25.png">

**Reply by mtiutiu-heits** [reply]
I see. Did you look at the volume in the DO panel ? Was it resized ? 

I'm testing the resize right now via Helm, and indeed the PVC size is not changed for some reason. I noticed that volume expansion is set for DO block storage, so if the PVC value is increased the DO block storage volume should be increased as well:

```text
$ kubectl get sc

NAME                         PROVISIONER                 RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
do-block-storage (default)   dobs.csi.digitalocean.com   Delete          Immediate           true                   10d
```

So, after I changed the volume size in the helm values file, I see the new value reflected in the output (`10Gi`):

```text
$ helm get values kube-prom-stack -n monitoring

...
prometheus:
  enabled: true
  prometheusSpec:
    storageSpec:
      volumeClaimTemplate:
        spec:
          accessModes:
          - ReadWriteOnce
          resources:
            requests:
              storage: 10Gi
          storageClassName: do-block-storage
...
```

Then, if I list the Prometheus PVC parameters, it still shows the old size of `5Gi`:

```text
$ kubectl get pvc -n monitoring

NAME                                                                                                     STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS       AGE
prometheus-kube-prom-stack-kube-prome-prometheus-db-prometheus-kube-prom-stack-kube-prome-prometheus-0   Bound    pvc-150c7961-18ec-497f-8a3e-bb5e0d29299c   5Gi        RWO            do-block-storage   88m
```

For me it seems that for some reason the Prometheus operator doesn't see the new storage spec value. 

Still investigating.

**Reply by mtiutiu-heits** [reply]
@ShurikAg 

Seems there's a [discussion](https://github.com/prometheus-operator/prometheus-operator/issues/4079) about this issue on the prometheus operator page. You can check that as well.

**Reply by mtiutiu-heits** [reply]
@ShurikAg 

In a nutshell, it seems that the Prometheus operator is  using StatefulSets under the hood. StatefulSets are immutable in general, and changing PVC size attached to them is subject to same restrictions.

After a quick search I found this [blog post](https://itnext.io/resizing-statefulset-persistent-volumes-with-zero-downtime-916ebc65b1d4) which explains the same thing.

Let me know if it's the same issue, and if it helps.

Thanks.

**Reply by ShurikAg** [reply]
@mtiutiu-heits  OK, that starts making a bit more sense. However, as I mentioned before, I accidentally deleted the storage class and I don't have the config that it used to have. Is there any way to resurrect it?

**Reply by ShurikAg** [reply]
@mtiutiu-heits  Thank you for pointing out that article. It was super helpful for information. However, it didn't work as described. I had to delete everything related to Prometheus, except the operator, to fix it. Not a big deal in our case, we didn't have downtime for the application. Also, since we are managing everything through Flux CD, recreating everything was as simple as committing new config with changes.
